flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java

```diff
@@ -68,20 +68,29 @@
 
 	private final boolean flushAlways;
 
+	private final boolean isBroadcastSelector;
+
 	private Counter numBytesOut = new SimpleCounter();
 
 	private Counter numBuffersOut = new SimpleCounter();
 
 	public RecordWriter(ResultPartitionWriter writer) {
-		this(writer, new RoundRobinChannelSelector<T>());
+		this(writer, new RoundRobinChannelSelector<T>(), false);
 	}
 
-	@SuppressWarnings("unchecked")
-	public RecordWriter(ResultPartitionWriter writer, ChannelSelector<T> channelSelector) {
-		this(writer, channelSelector, false);
+	public RecordWriter(
```

Comment:
I'm not sure if I like the current approach of passing broadcasting `ChannelSelector` that is never being used and separate `boolean isBroadcastSelector` flag. It's redundant and can lead to errors (like passing non broadcasting selector with flag set to true or vice versa). I think it would be nice good to try out if this code can be refactored in one of the two ways:

1. Adding some method like `boolean ChannelSelector::isBroadcast()` and use it in the `RecordWriter` constructor
2.  Disallowing to pass both those parameters and changing the signature of this constructor to:
```
RecordWriter(ResultPartitionWriter writer, boolean isBroadcast)
```

Maybe there is also some other way around?

Reply:
To be honest, I am also a bit embarrassed for current implementation. It does exist the redundant and potential inconsistent issues as you mentioned.

The first way you mentioned can solve above issues if we can check whether the selector is broadcast or not internally in `RecordWriter`.

For the second way we can not only pass the boolean argument without selector, because once the boolean value is false, we should rely on the specific selector to get channels. So the selector has to be given explicitly, but the boolean can be got implicitly from the selector.

I also thought of another way before, that is defining another BroadcastRecordWriter for handling directly. Then the writer is distinguished by common `RecordWriter` and special `BroadcastRecordWriter`. In common `RecordWriter`, the selector is needed for routing channels, and for `BroadcastRecordWriter` it does not need the selector at all. To do so we may need further extract the common codes for these two writers, furthermore we may need to remove current `StreamRecordWriter` which would be integrated with current `RecordWriter` directly. So the writer is divided by broadcast mode, not by stream mode. What do you think this way?

I would also further consider the changes for above two options through.

Reply:
Ops, right 2nd option was pretty stupid on my part 😳 

Isn't having separate optimised `BroadcastRecordWriter` our goal here and whole point of your interest in this area? From this perspective won't this if check:
```
 	public void emit(T record) throws IOException, InterruptedException {
 		if (isBroadcastSelector) {
 			broadcastEmit(record);
 		} else {
 			emit(record, channelSelector.selectChannels(record));
 		}
 	}
```
become a dead code/disappear shortly?

In that case indeed, maybe indeed the best way to solve this issue it so actually introduce simple `BroadcastRecordWriter` (which would extend `RecordWriter` and overwrite only one method
```
	@Override 
	public void emit(T record) throws IOException, InterruptedException {
		broadcastEmit(record, broadcastChannels);
	}
```
The only tricky part would be that that would require us to create either some builder method (or even `RecordWriterBuilder` class) something like:
```
public static RecordWriter createRecordWriter(ResultPartitionWriter writer, ChannelSelector<T> channelSelector) {
  // might return either BroadcastRecordWriter or RecordWriter based on passed channelSelector
}
```
?

After that the next step would be to optimize `BroadcastRecordWriter`.

What do you think? Does it make sense or have I missed something again 😳?

Reply:
Yes, my previous thought is very similar with your above demo.

The `BroadcastRecordWriter` can extend `RecordWriter` and optimize the emit path specially, then it can avoid `if` check for each emit.  And it might seem not very tricky to create `RecordWriter` or `BroadcastRecordWriter` based on `ChannelSelector` instance.

I can try this way for updating the codes later. :)

Reply:
I am considering the way for solving this issue.

I agree it seems better to introduce either a builder method or `RecordWriterBuilder` class for creating the specific RecordWriter. But it seems unfeasible to check ChannelSelector instance in current architecture, because the classes of `BatchTask` and `StreamTask` are in different modules and the` instanceof BroadcastPartitioner` check would only work under streaming module, not meanwhile work for `BatchTask` in runtime module.

There might be two ways for this issue:
1. Introduce `ChannelSelector#isBroadcast` method to solve the module dependence.
2. Check the broadcast mode separately in `StreamTask` and `BatchTask` classes, that means no unified builder for creating `RecordWriter`

I prefer the second way a bit in order not to change the interface much, although this check instance would be scattered in two different places. What do you think?

Reply:
`instanceof BroadcastPartitioner` sounds like a hack, not a proper solution. In the short term I would vote for `ChannelSelector#isBroadcast`. 

Maybe in the long term a better solution would be to introduce some kind of `ChannelSelectorBuilder` or `ChannelSelectorDescriptor`, which is used to create correct `RecordWriter` and `ChannelSelector`. It could also solve the issue of `BroadcastPartitioner#selectChannel()` throwing `UnsupportedOperationException` which is also ugly :( ). But I'm not sure if that's the best solution and I don't think it's worth doing yet another refactor here now.

Reply:
Yes, from the long time we might do not need the `BroadcastPartitioner` implementation because we do not need any specific methods from this partitioner, then it can avoid the current `UnsupportedOperationException` ugly way.

But it might refactor the implementation from DataStream path related to generate `BroadcastPartitioner`, and I would take the simple `ChannelSelector#isBroadcast` currently. :)